### PR TITLE
added stl.natvis to stl

### LINKED ISSUE #349: STL: Include stl.natvis in this repository
Natvis file should be in that repository and linked to pdb file ([/natvis](https://docs.microsoft.com/en-us/cpp/build/reference/natvis-add-natvis-to-pdb?view=vs-2019))
Example:
https://github.com/microsoft/GSL/blob/7e99e76c9761d0d0b0848b91f8648830670ee872/CMakeLists.txt#L79-L81

Also, `natstepfilter` and `natjmc` file could be also useful

### PR DESCRIPTION
**Pull request by AnjuDel**
Fixes #349.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -362,6 +362,16 @@ The `SKIPPED` result code indicates that a given test was explicitly skipped by
...
 
+# Editing And Testing The Debugger Visualizer
+
+### Modify The Visualizer
+
+To modify how elements look on the debugger open the file STL\stl\STL.natvis. For more information on how to modify this file check the [natvis documantation][].
+
+### Test your changes.
+
+After making your changes overwrithe the file under %Visual Studio instalation path%\Common7\Packages\Debugger\Visualizers\stl.natvis (is remcomended to make a backup of the file). After overwriting the file you should be able to see your chanes in a Visual Studio debugging sesion.
```
Comment (IhateTrains) [inline README.md:373]:
```suggestion
After making your changes overwrite the file under %Visual Studio instalation path%\Common7\Packages\Debugger\Visualizers\stl.natvis (is recommended to make a backup of the file). After overwriting the file you should be able to see your changes in a Visual Studio debugging session.

```

File: stl/STL.natvis [hunk 2]
```diff
@@ -0,0 +1,1596 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<AutoVisualizer xmlns="http://schemas.microsoft.com/vstudio/debugger/natvis/2010">
```
Comment (MahmoudGSaleh) [inline stl/STL.natvis:3]: Is 2010 the latest version? It's a good opportunity to update this file to the latest.

File: stl/STL.natvis [hunk 3]
```diff
@@ -0,0 +1,1596 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (MahmoudGSaleh) [inline stl/STL.natvis:1]: This doesn't affect the stl code or building, so I recommend moving it to a parallel folder. I recommend `<root>/debugger`.

File: README.md [hunk 4]
```diff
@@ -362,6 +362,16 @@ The `SKIPPED` result code indicates that a given test was explicitly skipped by
 * taking a very long time to run
 * failing or passing for the incorrect reason
 
+# Editing And Testing The Debugger Visualizer
+
+### Modify The Visualizer
+
+To modify how elements look on the debugger open the file STL\stl\STL.natvis. For more information on how to modify this file check the [natvis documantation][].
```
Comment (MahmoudGSaleh) [inline README.md:369]:
```suggestion
To modify how elements look in the debugger edit the file debugger\STL.natvis. For more information on how to modify this file check the [natvis documentation][].

```

File: README.md [hunk 5]
```diff
@@ -423,3 +433,4 @@ SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 [opencode@microsoft.com]: mailto:opencode@microsoft.com
 [redistributables]: https://support.microsoft.com/en-us/help/2977003/the-latest-supported-visual-c-downloads
 [vcpkg]: https://github.com/microsoft/vcpkg
+[natvis documantation]: https://docs.microsoft.com/en-us/visualstudio/debugger/create-custom-views-of-native-objects?view=vs-2019
```
Comment (MahmoudGSaleh) [inline README.md:436]:
```suggestion
[natvis documentation]: https://docs.microsoft.com/en-us/visualstudio/debugger/create-custom-views-of-native-objects?view=vs-2019

```

**Review by MahmoudGSaleh — CHANGES_REQUESTED** [review]
I've left a few comments. Mostly about the new file location and some spelling.

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: README.md [hunk 6]
```diff
@@ -362,6 +362,17 @@ The `SKIPPED` result code indicates that a given test was explicitly skipped by
 * taking a very long time to run
 * failing or passing for the incorrect reason
 
+# Editing And Testing The Debugger Visualizer
+
+### Modify The Visualizer
+
+To modify how components are visualized in the debugger edit the file STL\stl\STL.natvis. For more information on how to modify this file check the [natvis documentation][].
```
Comment (CaseyCarter) [inline README.md:369]: I think we have one too many "STL"s here; the file we're adding is named `stl\STL.natvis`.
```suggestion
To modify how components are visualized in the debugger edit the file stl\STL.natvis. For more information on how to modify this file check the [natvis documentation][].
```

File: stl/STL.natvis (hunk 3 above)
Comment (CaseyCarter) [reply stl/STL.natvis:1]: (FWIW, I'm happy with the now-current `stl/STL.natvis`.)

File: stl/STL.natvis [hunk 7]
```diff
@@ -0,0 +1,1596 @@
...
+  <Type Name="std::auto_ptr&lt;*&gt;">
+      <DisplayString Condition="_Myptr == 0">empty</DisplayString>
+      <DisplayString Condition="_Myptr != 0">auto_ptr {*_Myptr}</DisplayString>
+      <Expand>
+          <Item Condition="_Myptr != 0" Name="[ptr]">_Myptr</Item>
+      </Expand>
+  </Type>
+
+
+  <!--Visualizer for coroutines-->
```
Comment (CaseyCarter) [inline stl/STL.natvis:1571]: MSVC-PR-307511 is making changes here. Should we wait for that PR to land and merge those changes here, or flow those changes in a separate PR later?

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/STL.natvis (hunk 2 above)
Comment (AnjuDel) [reply stl/STL.natvis:3]: That file is the schema I don't think we need to update it.

File: stl/STL.natvis (hunk 7 above)
Comment (AnjuDel) [reply stl/STL.natvis:1571]: Fun indeed, I agree we should wait for that changes to go in

File: README.md (hunk 6 above)
Comment (CaseyCarter) [inline README.md:369]: This pathname is very wrong now:
```suggestion
To modify how components are visualized in the debugger edit the file stl\debugger\STL.natvis. For more information on how to modify this file check the [natvis documentation][].
```

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: README.md [hunk 8]
```diff
@@ -362,6 +362,17 @@ The `SKIPPED` result code indicates that a given test was explicitly skipped by
...
 
+# Editing And Testing The Debugger Visualizer
+
+### Modify The Visualizer
+
+To modify how components are visualized in the debugger edit the file STL\stl\STL.natvis. For more information on how to modify this file check the [natvis documentation][].
+
+### Test your changes.
+
+After making your changes copy the file to any c++ Visual Studio project, after that select your right click your project > Add > Existing Item and select the stl.natvis file. After doing this you should be able to see your changes in a Visual Studio debugging session.
```
Comment (SuperWig) [inline README.md:373]: > After making your changes copy the file to any c++ Visual Studio project

If you're using Add Existing Item why do you need to copy it?
```suggestion
You can add the natvis file to any Visual Studio C++ project if you right click your project > Add > Existing Item and select the stl.natvis file. After doing this you should be able to see your changes in a Visual Studio debugging session.
```

File: stl/STL.natvis (hunk 7 above)
Comment (CaseyCarter) [reply stl/STL.natvis]: I see the coroutine changes here but I think we're still missing [VS-PR-308324](https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/308324).

Comment (CaseyCarter) [reply stl/STL.natvis]: > I see the coroutine changes here but I think we're still missing [VS-PR-308324](https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/308324).

You are blind; I see both the `string_view` and `vector` changes here from that PR.

**Review by CaseyCarter — APPROVED** [review]

**Review by MahmoudGSaleh — APPROVED** [review]

File: README.md [hunk 9]
```diff
@@ -362,6 +362,20 @@ The `SKIPPED` result code indicates that a given test was explicitly skipped by
 * taking a very long time to run
 * failing or passing for the incorrect reason
 
+# Editing And Testing The Debugger Visualizer
+
+### Modify The Visualizer
+
+To modify how components are visualized in the debugger edit the file stl\debugger\STL.natvis. For more information on
```
Comment (StephanTLavavej) [inline README.md:369]: Nitpick: The readme generally uses code font for paths:
```suggestion
To modify how components are visualized in the debugger edit the file `stl\debugger\STL.natvis`. For more information on
```

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

**Review by CaseyCarter — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Marking as `blocked` to remind myself that we need a VS PR before merging this into GitHub and MSVC.

**Comment by StephanTLavavej** [comment]
@AnjuDel created VS-PR-311424 to mirror this. :tada:

**Comment by StephanTLavavej** [comment]
Thanks for adding the visualizers and writing up how to consume them - this is a significant step forward for the repo! :smiley_cat: :eye: :tada:

**Pull request merged**